learn/glossary.md

```diff
@@ -14,6 +14,15 @@ If you encounter a term you wish were defined here, please feel free to [file an
 
 The entries on this page can be linked to via anchor links (e.g. `https://json-schema.org/learn/glossary.html#vocabulary`) when sharing a definition with others.
 
+### dialect
+
+A collection of [vocabularies](#vocabulary), along with an indication of whether supporting each vocabulary is required to process schemas written in the dialect.
```

Comment:
I wonder if this definition is too tightly coupled to the Vocabulary System. First of all, the vocabulary system likely to change in the future possibly making this definition no longer make sense. Second, releases such as draft-07 that pre-date the Vocabulary System are considered dialects as well, but aren't defined by vocabularies. You can consider draft-07 as a single required vocabulary, but it's not defined that way.

I'd probably define a dialect as the set of keywords that are understood in a schema. Those keywords being defined by vocabularies is an artifact of the Vocabulary System and not necessarily a defining property of a dialect.

Reply:
Thanks! I took this definition basically with only minor change [from the spec](https://json-schema.org/draft/2020-12/json-schema-core.html#name-overview):

> A dialect is defined as a set of vocabularies and their required support identified in a meta-schema.

Obviously we can change this definition if/when we change it elsewhere, but are you still suggesting we do so beforehand? (Or have we already changed the definition elsewhere and I missed it?)

> Second, releases such as draft-07 that pre-date the Vocabulary System are considered dialects as well, but aren't defined by vocabularies.

Interesting -- do you have a source for this? I'm not doubting you, I've just not heard it before but possibly just wasn't paying attention well enough -- are you saying we should define dialects this way or that others already do? I guess it also occurred to me that this may be the case because we call `$schema` now a "dialect identifier" even though draft 7 did not have one, so retroactively we must call draft 7 a dialect, but equally well we could retroactively say the dialect is defined by a single vocabulary, no?

Reply:
> it also occurred to me that this may be the case because we call $schema now a "dialect identifier" even though draft 7 did not have one, so retroactively we must call draft 7 a dialect.

Correct.

> but equally well we could retroactively say the dialect is defined by a single vocabulary, no?

I don't know what this means.

Reply:
@Julian so does "reatroactively say the dialect is defined by a single vocabulary" refer to considering draft-07 to be a single vocabulary?  Austin started talking about multiple vocabularies all the way back in draft-wright-*-00 (a.k.a. not-draft-05) in 2016:

>    Other specifications [besides JSON Schema Core] define the vocabularies that perform assertions  about validation, linking, annotation, navigation, and interaction.

The concept of JSON Schema vocabularies had its own [section in the spec](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-00#section-4.3.2) in draft-07.
 
We just didn't settle on the exact number and granularity of them and _formalize_ vocabulary identification and selection until 2019-09.  But since 2016 we have always spoken of there being at least two vocabularies (core and validation) in the standard dialect, with hyper-schema as a third vocabulary.

Reply:
We didn't have vocabulary-level meta-schemas prior to 2019-09 because without `$recursive*` or `$dynamic*`, it was too much of a pain to manage.  Hyper-Schema was enough of a pain on its own.  But vocabulary-level meta-schemas are just conveniences and best practices, they have no formal role in anything as distinct entities.

Reply:
Got it, thanks, so you disagree (well, the spec you linked disagrees) with what Jason said then, i.e. 

> Second, releases such as draft-07 that pre-date the Vocabulary System are considered dialects as well, but aren't defined by vocabularies.

that's not true, they indeed are defined by vocabularies, in the form that existed at the time?

> @Julian so does "reatroactively say the dialect is defined by a single vocabulary" refer to considering draft-07 to be a single vocabulary?

(Yes -- I wasn't aware and didn't check that the section you linked was present in draft 7, I took Jason's word for it).

What about draft 4, which doesn't mention the word vocabulary, though it's otherwise fairly similar structurally to draft 7, is it a dialect, and/or is https://json-schema.org/draft-04/schema its dialect URI?

All I'm trying to do is interpret definitions (and claims here) that are written down into a definition.

Reply:
@Julian I think you're overthinking this a bit.  We weren't exactly trying to harmonize the terms across all past drafts.  Whether or not Jason's statement is true depends on whether he meant "vocabulary" in a general sense or "vocabulary" in a "think you can put in `$vocabulary`" sense, which feels like splitting hairs to me.

Meta-schema URIs are (retroactively) dialect URIs.  If you go back far enough (draft-04) a lot of things won't make sense, and we intentionally don't try to make it all make perfect sense.  We don't even want people using draft-04 anyway.

Reply:
What is the definition of a dialect? Is it different from what's here?

Reply:
I didn't think we had used the word "vocabulary" before 2019-09, but I @handrews's word for it. The thing I'm trying to avoid is confusion about the general concept of a vocabulary vs a vocabulary that's part of the vocabulary system. Readers should understand that this refers to the concept of vocabulary, not just the specific mechanism in the Vocabulary System.

Reply:
@Julian the term "dialect" emerged primarily over the course of a long series of discussions with OpenAPI, IIRC.  I don't remember if we had started using it during 2019-09, but it really became a thing between 2019-09 and 2020-12.  Of course, by the time 2020-12 went out I'd stepped away so I never got around to writing a formal definition into the spec.

I think I agree with @jdesrosiers that the concept of a vocabulary is more important than the mechanics, which are likely to change to some degree.  I haven't really thought about that in terms of how we want to define these words. 🤔